name: Double Nested Node Action - Pre-Post Conditional
on:
  workflow_dispatch:

jobs:
  double-nested-on-failure:
    runs-on: self-hosted
    continue-on-error: true
    steps:
      - run: exit 1
      - uses: fhammerl/canary/double-nested-node-pre-post-on-failure@main
        if: always()

  double-nested-on-failure-check:
    runs-on: self-hosted
    needs: double-nested-on-failure
    steps:
      - uses: actions/github-script@v5
        with:
          script: |
            const run_id = ${{ github.run_id }};
            async function getSuccessJobIdFromRunId(run_id) {
                return await github.request('GET /repos/{owner}/{repo}/actions/runs/{run_id}/jobs', {
                    owner: 'fhammerl',
                    repo: 'canary',
                    run_id
                })
            }

            async function getJobLogs(job_id) {
                return await github.rest.actions.downloadJobLogsForWorkflowRun({
                    owner: 'fhammerl',
                    repo: 'canary',
                    job_id
                })
            }

            getSuccessJobIdFromRunId(run_id).then(response => {
                const job_id = response.data.jobs.filter(job => job.name == "double-nested-on-failure")[0].id;
                getJobLogs(job_id).then(logs => {
                    if (logs.data.split(/\r?\n/).filter(line => line.indexOf("PRE-STEP-COMPLETE") != -1).length != 0) {
                        throw 'Pre-step ran but job is failed later on step 1'
                    }
                    if (logs.data.split(/\r?\n/).filter(line => line.indexOf("MAIN-STEP-COMPLETE") != -1).length == 0) {
                        throw 'Main step did not run!'
                    }
                    if (logs.data.split(/\r?\n/).filter(line => line.indexOf("POST-STEP-COMPLETE") != -1).length == 0) {
                        throw 'Post-step did not run!'
                    }

                    console.log("All checks successful!")
                });
            });

      - run: echo 'in case of actual failure do slack webhook' && exit 1
        id: todo-slack-webhook
        if: failure()
